#Section 1: Foundational Concepts

##1 Purpose of data manipulation in the data science workflow


Increasing the quality of your data by handling null value, removing unnecessary attribures and extracting certain features to enhance the accuracy and utility of your models during training

##2 Python Libraries used for data manipulation
1. Numpy
2. Pyspark
3. Scikit*

##3 Key limitations of python lists when performing computations with Large datasets
- Can only handle 2 dimensional data
- Its inbuilt funtions don't intuituitively support numerical operations
- Since its interpreted rather than compiled, it is significantly slower on larger datasets

##4 Difference between how python lists and numpy Arrays handle addition of two lists/arrays

- In python lists: adding two arrays concatanates them increasing the length of the list
- in numpy arrays: adding two arrays adds the corresponding elements in the similar indexed position(meaning you can only add arrays which are of the same order)

##5 An ndarray

This is a matrix of datapoints which can initialized to any number of dimensions in the numpy library

##6 Advantages of using Numpy arrays over Python lists for numberical computations

- Since numpy is written in C, numerical operations are much faster since the code is compiled
- Can support a variable number of dimensional data
- Operations are intuitive and written with numerical operations in mind making code much easier to write and understand

##7 Primary data structures Pandas introduces for working with tabular data and a brief description of each

- Series: This is a one dimensional data structure with a label similar to how a single column in excel might look like
- DataFrame: This is 2 dimensional data structure which allows for different different data types in each column. You can also lable each column with a specific index for quick reference.

##8 How Pandas simply working with tabular data compared to what Numpy primarily offers

Because dataframe allow for different types of datatypes in different columns and you can index each column with a particular label rather than the default integer index, this allows for intuitive transfer of data from excel sheets to python whereas ndarrays only support a single data type in each ndarray and it less easier to refference each column independenlty.

##9 Ways of accessing a column in a Pandas DataFrame with examples

You use the the [] or .method way. Assume we have a dataframe initialzed as df with a column called age we can access it by
- df.age
- df['age']

##10 What df.isnull().sum() tells us about a Dataframe in pandas and its importance in data cleaning
The total number of null data points per column in our entire dataset

#Section 2: Foundational Data Exploration and Manipulation

In [2]:
#initializing the dataframe and importing pandas
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep='\t')

##1. Displaying last 7 rows of the DataFrame

In [ ]:
print(df.tail(7))

      order_id  quantity            item_name  \
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   


##2. Summary of the DataFrame

In [ ]:
df.describe()

order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [ ]:
df.isnull().sum()

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

###comments
- The dataset has 5 columns and two of them are interger that is order_id an quantity. It has a total of 4622 rows
- Only choice_description has null values(1246 in total) and the data seems to be a transaction table from hotel database.
- Customers in this hotel averagely order a single of each menu item at a go with the highest quantity of a menu item having being 15
- The dataset is using 180.7 Kb of memory for storage, quite small and can be manipulated on most modern computers.

##3. Observations with missing item_price

In [1]:
df.item_price.isnull().sum()

NameError: name 'df' is not defined

all rows have non-null item_price

##4. Names of all columns in the dataframe

In [ ]:
print(df.columns)

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')


##5. Item_name which appears most frequently in the orders

In [ ]:
print(df['item_name'].value_counts().idxmax())

Chicken Bowl


- valuecounts method returns vales of each unique item price and their corresponing number of occurence. It is be default sorted in descending order and excludes the null values
- idxmax method returns the index of the maximum value(idx max) from a list of values with their corresponding counts

##6. Number of distinct itemnames present in the dataset

In [10]:
print(df.item_name.nunique())

50


- unique() method in Pandas returns the number of distinct items in a series while nunique returns the number of unique items

##7. New dataframe containing only "Chicken_bowl" in item name

In [13]:
df1 = df[df['item_name'] == 'Chicken Bowl']
print(df1)

      order_id  quantity     item_name  \
4            2         2  Chicken Bowl   
5            3         1  Chicken Bowl   
13           7         1  Chicken Bowl   
19          10         1  Chicken Bowl   
26          13         1  Chicken Bowl   
...        ...       ...           ...   
4590      1825         1  Chicken Bowl   
4591      1825         1  Chicken Bowl   
4595      1826         1  Chicken Bowl   
4599      1827         1  Chicken Bowl   
4604      1828         1  Chicken Bowl   

                                     choice_description item_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
19    [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75   
26    [Roasted Chili Corn Salsa (Medium), [Pinto Bea...     $8.49   
...                                                 ...        ...  
4590  [Roast

##8. Converting item_price to float value

In [24]:
df.item_price = df.item_price.apply(lambda x: x[1:]).astype(float)

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

- apply() function passes the specified funtion to each item in the series hence allowing removal of the $ sign and conversion to float

##9. Average Itemprice

In [29]:
df.item_price.mean()

np.float64(7.464335785374297)

##10. Total number of unique items based on order Id Column

In [30]:
df.order_id.nunique()

1834